wasm/src/logging.rs

```diff
@@ -0,0 +1,168 @@
+use core::sync::atomic::{AtomicU8, Ordering};
+
+use super::*;
+
+// NOTE: `u8::MAX` is a sentinel value for an undefined log level
+static MAX_LOG_LEVEL: MaxLogLevel = MaxLogLevel(AtomicU8::new(u8::MAX));
+
+/// Struct which holds a valid [`LogLevel`] stored as an integer
+struct MaxLogLevel(AtomicU8);
+
+/// Log level supported by the host
+// NOTE: This struct must be exact duplicate of `config::logger::Level`
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::upper_case_acronyms)]
+#[repr(u8)]
+pub enum LogLevel {
+    /// Trace
+    TRACE,
+    /// Debug
+    DEBUG,
+    /// Info (Default)
+    #[default]
+    INFO,
+    /// Warn
+    WARN,
+    /// Error
+    ERROR,
+}
+
+impl MaxLogLevel {
+    fn get(&self) -> LogLevel {
+        let mut log_level = self.0.load(Ordering::Relaxed);
+
+        if log_level == u8::MAX {
+            log_level = query_max_log_level() as u8;
+            self.0.store(log_level, Ordering::SeqCst);
+        }
+
+        // SAFETY: `MaxLogLevel` guarantees that transmute is valid
+        unsafe { core::mem::transmute(log_level) }
+    }
+}
+
+#[cfg(not(test))]
+mod host {
+    #[link(wasm_import_module = "iroha")]
+    extern "C" {
+        /// Get the max log level set on the host
+        pub(super) fn query_max_log_level() -> u32;
+
+        /// Log string with the host logging system
+        ///
+        /// # Warning
+        ///
+        /// This function doesn't take ownership of the provided allocation
+        pub(crate) fn log(log_level: u32, ptr: *const u8, len: usize);
+    }
+}
+
+/// Query the max log level set on the host
+fn query_max_log_level() -> LogLevel {
+    #[cfg(not(test))]
+    use host::query_max_log_level as host_query_max_log_level;
+    #[cfg(test)]
+    use tests::_query_max_log_level_mock as host_query_max_log_level;
+
+    // SAFETY: `host_query_max_log_level` must return a valid log level
+    #[allow(clippy::cast_possible_truncation)]
+    unsafe {
+        core::mem::transmute(host_query_max_log_level() as u8)
+    }
+}
+
+/// Log `obj` with desired log level
+pub fn log<T: Encode + ?Sized>(log_level: LogLevel, obj: &T) {
+    #[cfg(not(test))]
+    use host::log as host_log;
+    #[cfg(test)]
+    use tests::_log_mock as host_log;
+
+    if log_level >= MAX_LOG_LEVEL.get() {
```

Comment:
Currently smart-contract won't change it'sloglevel if host change it's loglevel, however i don't think it's significant problem.